Create an indexed task and find it smoketest

Rolikasi opened the pull request:
Add smoketest that creates indexed task and finds it in the index
Closes #1244

Rolikasi commented:
Is it ok that these smoketest indexed tasks expire in a year?

djmitche commented:
> Is it ok that these smoketest indexed tasks expire in a year?

An hour would be better :)

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,48 @@
+  title: 'create an indexed task and find it in the index',
+  requires: [],
+  provides: [
+    'target-indexTask',
```
let's use `-` here, so `target-indexed-task`

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,48 @@
+    utils.status({message: 'indexTask-find taskId: ' + randomId});
+    await queue.createTask(randomId, task);
+    let index = new taskcluster.Index(taskcluster.fromEnvVars());
+    index.findTask(findIndexedTask);
```
This call must be `await`ed

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,48 @@
+        });
+        await new Promise(resolve => setTimeout(resolve, 1000));
+      } else if (status.status.state === "completed") {
+        return;
```
I think we would want to check here that the task is in the index?

Note that the task is indexed *after* it completes, so it may be necessary to wait another few seconds before `findTask` finds the task.

Rolikasi commented:
@djmitche kindly review the changes) I think it's checking now that the task is in the index)

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,50 @@
+const taskcluster = require('taskcluster-client');
+
+exports.tasks = [];
```
If you rebase onto the latest master, you'll see this needs a `scopeExpression` giving the scopes it requires.

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,50 @@
+        source: "https://taskcluster.net",
+      },
+      payload: {},
+      routes: [`index.project.taskcluster.smoketest.${randomId}`],
```
```suggestion
      routes: [`index.${taskIndex}`],
```

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,50 @@
+  run: async (requirements, utils) => {
+    let queue = new taskcluster.Queue(taskcluster.fromEnvVars());
+    let randomId = taskcluster.slugid();
+    const findIndexedTask='project.taskcluster.smoketest.' + randomId;
```
maybe call this variable `taskIndex`?

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,50 @@
+      } else if (finded) {
+        return;
+      } else {
+        throw new Error('Task finished with status ' + status.status.state);
```
I think we'll end up here if we happen to poll *after* the task has completed but before the index service has inserted it into the index.  That's not a failure case -- it's a case where we should wait 1s and try again.

Rolikasi commented:
@djmitche kindly review the changes)

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,60 @@
+        return;
+      } else {
+        await new Promise(resolve => setTimeout(resolve, 1000));
+      }
```
This conditional could be refactored a little by moving the sleep outside of the conditional, since it happens in two of the three branches of control flow, while the third branch returns.

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,60 @@
+    let pollForStatusStart = new Date();
+    while((new Date() - pollForStatusStart) < 120000){
+      let status = await queue.status(randomId);
+      let finded = await index.findTask(taskIndex);
```
"found" is the typical English word.

Rolikasi replied:
oups

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,60 @@
+    utils.status({message: 'indexTask-find taskId: ' + randomId});
+    await queue.createTask(randomId, task);
+    let index = new taskcluster.Index(taskcluster.fromEnvVars());
+    await new Promise(resolve => setTimeout(resolve, 5000));
```
I commented this out, and the test failed.  That means that something in this test is dependent on timing, and that often leads to intermittent test failures.  Can you remove this and fix the issue causing the test to fail?

Rolikasi commented:
After weekend update I guess there is no role that give me the necessary scopes to run the smoketests, am I right?

Rolikasi commented:
The problem was that I tried to find a task before its state was "completed". I made some changes now I think it works correctly

Rolikasi commented:
@djmitche kindly review the changes)

djmitche commented on infrastructure/tooling/src/smoketest/checks/indexTask.js:
```diff
@@ -0,0 +1,58 @@
+        await index.findTask(taskIndex);
+        return;
+      } else {
+        await new Promise(resolve => setTimeout(resolve, 1000));
```
This will only happen in the case where the task is not pending or running, meaning that the check will busy-loop without any sleeping while the task is pending or running.

Rolikasi commented:
Thank for the review @djmitche) 
Added try statement to prevent errors when task could not be indexed right after it's completed
Moved sleep out of the condition
Deleted else condition
